### PR TITLE
Update Set-DbaDbQueryStoreOption.ps1

### LINKED ISSUE #9203: Set-DbaDbQueryStoreOption fails if database snapshots exist and aren't explicitly excluded
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

```
Exception calling "Alter" with "0" argument(s): "Alter failed for  'Microsoft.SqlServer.Management.Smo.ObjectKeyBase'. "
At line:75358 char:25
+                         $db.QueryStoreOptions.Alter()
+                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : FailedOperationException
 
WARNING: [16:37:57][Set-DbaDbQueryStoreOption] Could not modify configuration. | The operation cannot be performed on a database snapshot.
ALTER DATABASE statement failed.


writeErrorStream      : True
PSMessageDetails      : 
Exception             : System.Exception: The operation cannot be performed on a database snapshot.
                        ALTER DATABASE statement failed. ---> Microsoft.Data.SqlClient.SqlException: The operation cannot be performed on a database snapshot.
                        ALTER DATABASE statement failed.
                           at Microsoft.SqlServer.Management.Common.ConnectionManager.ExecuteTSql(ExecuteTSqlAction action, Object execObject, DataSet fillDataSet, Boolean catchException)
                           at Microsoft.SqlServer.Management.Common.ServerConnection.ExecuteNonQuery(String sqlCommand, ExecutionTypes executionType, Boolean retry)
                           --- End of inner exception stack trace ---
TargetObject          : [**db_name_redacted**]
CategoryInfo          : InvalidOperation: ([**db_name_redacted**]:PSObject) [Write-Error], Exception
FullyQualifiedErrorId : dbatools_Set-DbaDbQueryStoreOption,Stop-Function
ErrorDetails          : The operation cannot be performed on a database snapshot.
                        ALTER DATABASE statement failed.
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Stop-Function, <No file>: line 97916
                        at Set-DbaDbQueryStoreOption<Process>, <No file>: line 75460
                        at <ScriptBlock>, <No file>: line 7
PipelineIterationInfo : {0, 1}
```

### Steps to Reproduce

Create a database snapshot of any database on an instance - this can only be done via a T-SQL command rather than through the GUI of SQL Management Studio. See the [BOL ](https://learn.microsoft.com/en-us/sql/relational-databases/databases/create-a-database-snapshot-transact-sql?view=sql-server-ver16) for pre-requisites and details of the command.

```sql
CREATE DATABASE database_snapshot_name
ON
(
    NAME = logical_file_name,
    FILENAME = 'os_file_name'
)
AS SNAPSHOT OF source_database_name
```
Now try to standardise your Query Store settings for the instance by using `Set-DbaDbQueryStoreOption` without specifically excluding this new snapshot.

```powershell
Set-DbaDbQueryStoreOption -SqlInstance "**instance_name_redacted**" `
                -State ReadWrite `
                -StaleQueryThreshold 90 `
                -FlushInterval 900 `
                -CollectionInterval 30 `
                -MaxSize 500 `
                -MaxPlansPerQuery 200 `
                -WaitStatsCaptureMode On
```

Whilst it's relatively simple to pass in an exclusion list that includes any snapshots, it would be simpler if the command handled this as a known limitation.

### Please confirm that you are running the most recent version of dbatools

2.1.6

### Other details or mentions

_No response_

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe), Windows PowerShell ISE (powershell_ise.exe)

### PowerShell Host Version

Name                           Value                                                                                                                                                                                                                
----                           -----                                                                                                                                                                                                                
PSVersion                      5.1.19041.3803                                                                                                                                                                                                       
PSEdition                      Desktop                                                                                                                                                                                                              
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                              
BuildVersion                   10.0.19041.3803                                                                                                                                                                                                      
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                      
WSManStackVersion              3.0                                                                                                                                                                                                                  
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                  
SerializationVersion           1.1.0.1                                                                                                                                                                                                              




### SQL Server Edition and Build number

Microsoft SQL Server 2019 (RTM-CU22-GDR) (KB5029378) - 15.0.4326.1 (X64) 
	Aug 18 2023 14:05:15 
	Copyright (C) 2019 Microsoft Corporation
	Developer Edition (64-bit) on Windows 10 Enterprise 10.0 <X64> (Build 19045: ) (Hypervisor)


### .NET Framework Version

.NET Framework 4.8.9195.0

### PR DESCRIPTION
**Pull request by tominyorks**
Ignore snapshots when setting Query Store options

Set-DbaDbQueryStoreOptions uses ALTER DATABASE statements under the hood, and these fail when run against a read only database snapshot. Update the call to Get-DbaDatabase to exclude snapshots upfront.

## Please read -- recent changes to our repo
On November 10, 2022, [we removed some bloat from our repository (for the second and final time)](https://github.com/dataplat/dbatools/issues/8542). This change requires that all contributors reclone or refork their repo.

PRs from repos that have not been recently reforked or recloned will be closed and @potatoqualitee will cherry-pick your commits and open a new PR with your changes.

 - [x] Please confirm you have the smaller repo (85MB .git directory vs 275MB or 110MB or 185MB .git directory)

## Type of Change

 - [x] Bug fix (non-breaking change, fixes #9203 )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (affects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose
Stop Set-DbaDbQueryStoreOption from failing when a database snapshot exists but isn't explicitly added to an exclusion list

### Approach
Excludes database snapshots when retrieving the available databases from an instance

### Commands to test
`Set-DbaDbQueryStoreOption` is the only command affected.

Running the below, targeting a SQL instance that hosts at least one database snapshot will no longer error in the fixed version. Note, this will enable Query Store on **all** databases unless the `-Database` parameter is also provided.

```
Set-DbaDbQueryStoreOption -SqlInstance localhost `
    -State ReadWrite `
    -StaleQueryThreshold 90 `
    -FlushInterval 900 `
    -CollectionInterval 30 `
    -MaxSize 500 `
    -MaxPlansPerQuery 200 `
    -WaitStatsCaptureMode On
```

### CODE REVIEW

File: public/Set-DbaDbQueryStoreOption.ps1 [hunk 1]
```diff
@@ -170,7 +170,7 @@ function Set-DbaDbQueryStoreOption {
             }
 
             # We have to exclude all the system databases since they cannot have the Query Store feature enabled
-            $dbs = Get-DbaDatabase -SqlInstance $server -ExcludeDatabase $ExcludeDatabase -Database $Database | Where-Object IsAccessible
+            $dbs = Get-DbaDatabase -SqlInstance $server -ExcludeDatabase $ExcludeDatabase -Database $Database | Where-Object { ($_.IsAccessible) -and (!$_.IsDatabaseSnapshot) }
```
Comment (niphlod) [inline public/Set-DbaDbQueryStoreOption.ps1:173]: does `{ $_.IsAccessible -and !$_.IsDatabaseSnapshot }` work, too ? If so, it's a little bit clearer.

Comment (niphlod) [reply public/Set-DbaDbQueryStoreOption.ps1:173]: even better, remove the filter alltogether and use it like on https://github.com/dataplat/dbatools/blob/6610b1190e6dbffca49124ba7425c5ad95ed0ff6/public/Set-DbaDbQueryStoreOption.ps1#L178C17-L181C18 to signal the db is a snapshot, skipping it

Comment (tominyorks) [reply public/Set-DbaDbQueryStoreOption.ps1:173]: > does `{ $_.IsAccessible -and !$_.IsDatabaseSnapshot }` work, too ? If so, it's a little bit clearer.

It does - and it goes along with the conventions I usual ly follow too so not sure why I didn't here!

> even better, remove the filter alltogether and use it like on https://github.com/dataplat/dbatools/blob/6610b1190e6dbffca49124ba7425c5ad95ed0ff6/public/Set-DbaDbQueryStoreOption.ps1#L178C17-L181C18 to signal the db is a snapshot, skipping it

I'd resprectfully disagree this is better for this particular check. The existing check for `IsAccessible` makes sense - you'd want to make sure that something that was accessible when extracting the list of databases is still accessible now we want to change it. Snapshots are always snapshots though, they're not going to suddenly change into an actual database that we can configure.

Comment (niphlod) [reply public/Set-DbaDbQueryStoreOption.ps1]: makes sense, thanks

**Review by niphlod — APPROVED** [review]

**Comment by potatoqualitee** [comment]
hey nice, thank you both 💯 !

**Pull request merged**